cpp/docs/TESTING.md

```diff
@@ -0,0 +1,447 @@
+# Unit Testing in libcudf
+
+Unit tests in libcudf are written using 
+[Google Test](https://github.com/google/googletest/blob/master/googletest/docs/primer.md).
+
+**Important:** Instead of including `gtest/gtest.h` directly, use 
+`#include <cudf_test/cudf_gtest.hpp>`.
+
+## Best Practices: What Should We Test?
+
+In general we should test to make sure all code paths are covered. This is not always easy or 
+possible. But generally this means we test all supported combinations of algorithms and data types,
+and all operators supported by algorithms that support multiple operators (e.g. reductions, 
+groupby).  Here are some other guidelines.
+
+ * In general empty input is not an error in libcudf. Typically empty input results in empty output.
+   Tests should verify this.
+
+ * Anything that involves manipulating bitmasks (especially hand-rolled kernels) should have tests 
+   that check varying number of rows, especially around boundaries like the warp size (32). So, test
+   fewer than 32 rows, more than 32 rows, exactly 32 rows, and greater than 64 rows.
+
+ * Most algorithms should have one or more tests exercising inputs with a large enough number of 
+   rows to require launching multiple thread blocks, especially when values are ultimately 
+   communicated between blocks (e.g. reductions). This is especially important for custom kernels 
+   but also applies to Thrust and CUB algorithm calls with lambdas / functors. 
+
+ * For anything involving strings or lists, test exhaustive combinations of empty strings/lists,
+   null strings/lists and strings/lists with null elements.
```

Comment:
This came up today in a review.

"Also, for strings tests, include a mixture of non-ASCII UTF-8 characters like é in your test strings."